<?xml version="1.0" encoding="UTF-8"?>
<robot name="ur5_rg2">

    <!-- Fix to world-->
    <joint name="ur5_rg2_joint_world" type="fixed">
        <parent link="Base"/>
        <child link="base_link"/>
    </joint>

    <link name="Base">
    </link>

    <link name="base_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/base.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="shoulder_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/shoulder.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="upper_arm_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/upperarm.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="forearm_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/forearm.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="wrist_1_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/wrist1.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="wrist_2_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/wrist2.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="wrist_3_link">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/ur5/wrist3.dae" scale="1 1 1"/>
            </geometry> 
        </visual>
    </link>


    <!-- Arm joints-->
    <joint name="shoulder_pan_joint" type="revolute">
        <origin xyz="0.0 0.0 0.089159" rpy="0.0 0.0 0.0"/>
        <parent link="base_link"/>
        <child link="shoulder_link"/>
        <axis xyz="0.0 0.0 1"/>
        <limit lower="-6.28" upper="6.28" effort="150.0" velocity="3.15"/>
    </joint>

    <joint name="shoulder_lift_joint" type="revolute">
        <origin xyz="0.0 0.13585 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="shoulder_link"/>
        <child link="upper_arm_link"/>
        <axis xyz="0.0 1.0 0.0"/>
        <limit lower="-6.28" upper="6.28" effort="150.0" velocity="3.15"/>
    </joint>

    <joint name="elbow_joint" type="revolute">
        <origin xyz="0.0 -0.1197 0.425" rpy="0.0 0.0 0.0"/>
        <parent link="upper_arm_link"/>
        <child link="forearm_link"/>
        <axis xyz="0.0 1.0 0.0"/>
        <limit lower="-6.28" upper="6.28" effort="150.0" velocity="3.15"/>
    </joint>

    <joint name="wrist_1_joint" type="revolute">
        <origin xyz="0.0 0.0 0.39225" rpy="0.0 0.0 0.0"/>
        <parent link="forearm_link"/>
        <child link="wrist_1_link"/>
        <axis xyz="0.0 1.0 0.0"/>
        <limit lower="-6.28" upper="6.28" effort="28.0" velocity="3.2"/>
    </joint>

    <joint name="wrist_2_joint" type="revolute">
        <origin xyz="0.0 0.093 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="wrist_1_link"/>
        <child link="wrist_2_link"/>
        <axis xyz="0.0 0.0 1.0"/>
        <limit lower="-6.28" upper="6.28" effort="28.0" velocity="3.2"/>
    </joint>

    <joint name="wrist_3_joint" type="revolute">
        <origin xyz="0.0 0.0 0.09465" rpy="0.0 0.0 0.0"/>
        <parent link="wrist_2_link"/>
        <child link="wrist_3_link"/>
        <axis xyz="0.0 1.0 0.0"/>
        <limit lower="-6.28" upper="6.28" effort="28.0" velocity="3.2"/>
    </joint>



    <!-- Arm hand fixed joint-->
    <joint name="ur5_hand_joint" type="fixed">
        <origin xyz="0.0 0.0823 0.0" rpy="1.570796325 0.0 1.570796325"/>
        <parent link="wrist_3_link"/>
        <child link="rg2_hand"/>
    </joint> 

    <link name="rg2_hand">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/rg2/hand.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="rg2_leftfinger">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/rg2/finger.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <link name="rg2_rightfinger">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="3.141592653589793 0.0 0.0"/>
            <geometry>
                <mesh filename="package://project2/models/ur5_rg2/meshes/visual/rg2/finger.dae" scale="1 1 1"/>
            </geometry>
        </visual>
    </link>

    <!-- finger joints-->
    <joint name="rg2_finger_joint1" type="revolute">
        <origin xyz="0.105 0.017 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="rg2_hand"/>
        <child link="rg2_leftfinger"/>
        <axis xyz="0.0 0.0 1.0"/>
        <limit lower="0.0" upper="1.18" effort="10.6" velocity="1.57"/>
    </joint>

    <joint name="rg2_finger_joint2" type="revolute">
        <origin xyz="0.105 -0.017 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="rg2_hand"/>
        <child link="rg2_rightfinger"/>
        <limit lower="0.0" upper="1.18" effort="10.6" velocity="1.57"/>
        <axis xyz="0.0 0.0 -1.0"/>
    </joint>


    <!-- End effector-->
    <link name="tool0">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
    </link>
    <joint name="end_effector_frame_fixed_joint" type="fixed">
        <origin xyz="0.0 0.275 0.0" rpy="-1.570796325 0.0 0.0"/>
        <parent link="wrist_3_link"/>
        <child link="tool0"/>
    </joint>

    <!-- ====================== ROS 2 CONTROL ====================== -->
    <ros2_control name="GazeboSimSystem" type="system">
      <hardware>
        <plugin>gz_ros2_control/GazeboSimSystem</plugin>
      </hardware>

      <!-- 6DOF ramienia -->
      <joint name="shoulder_pan_joint">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="shoulder_lift_joint">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="elbow_joint">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="wrist_1_joint">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="wrist_2_joint">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="wrist_3_joint">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <!-- chwytak RG2 (opcjonalnie sterowany osobnym kontrolerem) -->
      <joint name="rg2_finger_joint1">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      <joint name="rg2_finger_joint2">
        <command_interface name="position"/>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
    </ros2_control>
</robot>